python/bin/watchman-wait-aio

```diff
@@ -0,0 +1,222 @@
+
+    def formatField(self, fname, val):
+        if fname == 'name':
+            # Respect the --relative path printing option
+            return os.path.relpath(val,
+                    args.relative.encode('utf-8')).decode('utf-8')
+        # otherwise just make sure it's a string so that we can join it
+        return str(val)
+
+    async def process(self, client):
+        global total_events
+        data = await client.get_subscription(self.name, self.root)
+        if data is None:
+            return
+        for f in data['files']:
+            out = []
+            if len(args.fields) == 1:
+                # When only 1 field is specified, the result is a
+                # list of just the values
+                out.append(self.formatField(args.fields[0], f))
+            else:
+                # Otherwise it is a list of objects
+                for fname in args.fields:
+                    out.append(self.formatField(fname, f[fname]))
+            print(args.separator.join(out))
+            sys.stdout.flush()
+            total_events = total_events + 1
+            if args.max_events > 0 and total_events >= args.max_events:
+                sys.exit(0)
+
+    async def process_loop(self, client):
+        while True:
+            await self.process(client)
+
+    @classmethod
+    def from_path(cls, path):
+        if path in subscriptions:
+            raise ValueError('path %s already specified' % path)
+        full_path = os.path.abspath(path)
+        if not os.path.exists(full_path):
+            print('''path %s (%s) does not exist.
+Perhaps you should use the --pattern option?''' % (path, full_path),
+                file=sys.stderr)
+            sys.exit(1)
+        subscriptions[path] = cls(path, full_path)
+
+
+def with_timeout(task):
+    if args.timeout > 0:
+        return asyncio.wait_for(task, args.timeout)
+    return task
+
+
+for path in args.path:
+    Subscription.from_path(path)
+
+loop = asyncio.get_event_loop()
+
+# Create the client and use the loop to open a connection to Watchman
+client = WatchmanClient(loop)
+
+# Schedule the packet receiving loop to run as an indefinite task.
+asyncio.ensure_future(client.receive_loop(), loop=loop)
+
+try:
+    # Check capabilities and start subscriptions synchronously
+    loop.run_until_complete(client.capability_check(
+        required=['term-dirname', 'cmd-watch-project', 'wildmatch']))
+    for _, sub in subscriptions.items():
+        loop.run_until_complete(with_timeout(sub.start(client)))
+
+    # Start the subscription processing tasks
+    tasks = [with_timeout(s.process_loop(client)) for _, s in
+            subscriptions.items()]
+    loop.run_until_complete(asyncio.gather(*tasks, loop=loop))
+
+except pywatchman.CommandError as ex:
+    print('watchman:', ex.msg, file=sys.stderr)
+    sys.exit(1)
+
+except KeyboardInterrupt:
+    # suppress ugly stack trace when they Ctrl-C
+    sys.exit(3)
```

Comment:
You'll still get warnings from asyncio about not closing the event loop and pending tasks. You should add a finally: block that cancels pending tasks, joins them and closes the loop.